File: android-lib/src/main/java/com/kinvey/android/KinveyHandlerThread.java
package com.kinvey.android;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;

import java.util.concurrent.ConcurrentLinkedQueue;

public class KinveyHandlerThread extends HandlerThread {

    private Handler mWorkerHandler;
    private ConcurrentLinkedQueue<Runnable> pendingQueue;

    public KinveyHandlerThread(String name, int priority) {
        super(name, priority);
        pendingQueue = new ConcurrentLinkedQueue<>();
    }

    public KinveyHandlerThread(String name) {
        super(name);
        pendingQueue = new ConcurrentLinkedQueue<>();
    }

    public synchronized void postTask(Runnable task){
        if (mWorkerHandler != null) {
            mWorkerHandler.post(task);
        } else {
            pendingQueue.add(task);
        }
    }

    @Override
    protected synchronized void onLooperPrepared() {
        mWorkerHandler = new Handler(getLooper());
        if (pendingQueue.size() > 0) {
            for (Runnable task:pendingQueue) {
                postTask(task);
            }
        }
    }

    synchronized void stopHandlerThread() {
        if (mWorkerHandler != null) {
            mWorkerHandler.post(new Runnable() {
                @Override
                public void run() {
                    Looper.myLooper().quit();
                    mWorkerHandler.removeCallbacksAndMessages(null);
                }
            });
        }
    }

}

File: android-lib/src/main/java/com/kinvey/android/KinveyHandlerThread.kt
package com.kinvey.android

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper

import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class KinveyHandlerThread : HandlerThread {

    private var mWorkerHandler: Handler? = null
    private var pendingQueue: ConcurrentLinkedQueue<Runnable>? = null
    private var executor: ExecutorService? = null

    constructor(name: String, priority: Int) : super(name, priority) {
        pendingQueue = ConcurrentLinkedQueue()
    }

    constructor(name: String) : super(name) {
        pendingQueue = ConcurrentLinkedQueue()
    }

    @Synchronized
    fun postTask(task: Runnable) {
        if (Client.sharedInstance.isClientRequestMultithreading) {
            if (executor == null) {
                executor = Executors.newFixedThreadPool(Client.sharedInstance.numberThreadPool)
            }
            executor?.submit(task)
        } else {
            if (mWorkerHandler != null) {
                mWorkerHandler?.post(task)
            } else {
                pendingQueue?.add(task)
            }
        }
    }

    @Synchronized
    override fun onLooperPrepared() {
        mWorkerHandler = Handler(looper)
        pendingQueue?.onEach { task -> postTask(task) }
    }

    @Synchronized
    internal fun stopHandlerThread() {
        if (mWorkerHandler != null) {
            mWorkerHandler?.post {
                Looper.myLooper()?.quit()
                mWorkerHandler?.removeCallbacksAndMessages(null)
            }
        }
    }

}
